netbsd/usr.sbin/cnwctl/cnwctl.8
.\" Copyright (c) 1997 Berkeley Software Design, Inc. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that this notice is retained,
.\" the conditions in the following notices are met, and terms applying
.\" to contributors in the following notices also apply to Berkeley
.\" Software Design, Inc.
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed by
.\"	Berkeley Software Design, Inc.
.\" 4. Neither the name of the Berkeley Software Design, Inc. nor the names
.\"    of its contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY BERKELEY SOFTWARE DESIGN, INC. ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL BERKELEY SOFTWARE DESIGN, INC. BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	BSDI $Id: cnwctl.8,v 1.1 1999/11/29 12:25:38 itojun Exp $
.Dd January 5, 1997
.Dt CNWCTL 8
.Os BSD/OS
.Sh NAME
.Nm cnwctl
.Nd display statistics and control Netwave AirSurfer PC Cards
.Sh SYNOPSIS
.Nm cnwctl
.Op Fl d Ar domain
.Op Fl i Ar interface
.Op Fl k Ar scramble-key
.Op Fl sS Op Ar rate
.Sh DESCRIPTION
The
.Nm cnwctl
utility is used to control Netwave AirSurfer PC Cards as well as
display statistics.
The following options are available:
.
.Bl -tag -width indent
.It Fl d
Set the domain of the card to
.Ar domain .
The domain must be between 0x000 and 0x1ff.
The domains 0x000 through 0x0ff are for access to an add-hoc network.
The domains 0x100 through 0x1ff are for access to a Netwave Access Point.
The default domain is 0x000.
A card may only talk to the access point in its domain, or other cards
in its add-hoc domain.
.It Fl i
Use
.Ar interface
as the interface rather than
.Li cnw0 .
.It Fl k
Set the scramble key to
.Ar scramble-key .
The scramble key must be between 0x0000 and 0xffff.
Both the source and the destination must use the same scramble key
in order to communicate.
.It Fl s
Display statistics.
When the optional argument
.Ar rate
(which must be the last argument to the end of the command line)
is specified as a non-zero value, statistics will be displayed
every
.Ar rate
seconds.
At the top of each
.Dq page
of statistics, column labels will be displayed.  The first row of
statistics will be totals since boot, subsequent lines are deltas
from the previous row.
If
.Ar rate
is not specified, or is 0 (zero), a single page of statistics will be
displayed.  These statistics are more detailed and include:
.Bl -tag -width indent
.It domain
The domain this card is part of.
.It rx
Number of packets received.
.It rxoverflow
Number of overflows detected.
.It rxoverrun
Number of overruns detected.
.It rxcrcerror
Number of CRC errors detected.
Random noise can cause these errors.
.It rxframe
Number of framing errors detected.
.It rxerrors
Number of generic errors detected.
.It rxavail
Number of times a packet was available.
.It tx
Number of packets requested to be sent.
.It txokay
Number of packets sent.
.It txabort
Number of packets aborted (not sent within 9 tries).
.It txlostcd
Number of times carrier detect was lost.
.It txerrors
Number of generic transmit errors detected.
.It txretries
Total number of retries.
.It Ar N Ns No x retries
Number of packets which were retried
.Ar N
times.
.El
.It Fl S
Display status read from the hardware.  This option is only available
to the super user.
The
.Ar rate
argument is used as with the
.Fl s
option.
The following fields are displayed
.Bl -tag -width indent
.It link integrity field (lif)
A 0 value implies no links.
.It connection quality (cq)
Probably indicates the quality of the connection to the access point.
.It spu
Unknown meaning.
.It link quality (lq)
Probably indicated the quality of the link to the access point.
.It hhc
Unknown meaning.
.It mhs
Unknown meaning.
.It revision
The revision numbers of the card.
.It id
The ID of the card.
.El
.El
.Sh SEE ALSO
.Xr cnw 4


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

netbsd/usr.sbin/cnwctl/cnwctl.c
/*
 * Copyright (c) 1997 Berkeley Software Design, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that this notice is retained,
 * the conditions in the following notices are met, and terms applying
 * to contributors in the following notices also apply to Berkeley
 * Software Design, Inc.
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by
 *	Berkeley Software Design, Inc.
 * 4. Neither the name of the Berkeley Software Design, Inc. nor the names
 *    of its contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY BERKELEY SOFTWARE DESIGN, INC. ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL BERKELEY SOFTWARE DESIGN, INC. BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/cdefs.h>
#include <sys/errno.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

#include <net/if.h>

#include <dev/pcmcia/if_cnwioctl.h>

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

void
main(int argc, char **argv)
{
	int c, domain, i, key, rate, sflag, Sflag, skt;
	char *e, *interface;
	struct ifreq ifr;
        struct cnwistats cnwis, onwis;
        struct cnwstatus cnws;
	struct ttysize ts;

	domain = -1;
	key = -1;
	Sflag = sflag = 0;
	rate = 0;
	interface = "cnw0";

	while ((c = getopt(argc, argv, "d:i:k:sS")) != EOF)
		switch (c) {
		case 'd':
			domain = strtol(optarg, &e, 0);
			if (e == optarg || *e || domain & ~ 0x1ff)
				errx(1, "%s: invalid domain", optarg);
			break;
		case 'i':
			interface = optarg;
			break;
		case 'k':
			key = strtol(optarg, &e, 0);
			if (e == optarg || *e || key & ~ 0xffff)
				errx(1, "%s: invalid scramble key", optarg);
			break;
		case 'S':
			++Sflag;
			break;
		case 's':
			++sflag;
			break;
		default: usage:
			fprintf(stderr, "usage: cnwctl [-i interface] [-d domain] [-k key] [-sS [rate]]\n");
			exit(1);
		}

	switch (argc - optind) {
	case 0:
		break;
	case 1:
		if (sflag == 0 && Sflag == 0)
			goto usage;
		if (sflag && Sflag)
			errx(1, "only one of -s and -S may be specified with a rate");
		rate = strtol(argv[optind], &e, 0);
		if (e == optarg || *e || rate < 1)
			errx(1, "%s: invalid rate", optarg);
		break;
	default:
		goto usage;
	}

        if ((skt = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
                err(1, "socket(AF_INET, SOCK_DGRAM)");

	if (key >= 0) {
		memset(&ifr, 0, sizeof(ifr));
		strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
		ifr.ifr_key = key;
		if (ioctl(skt, SIOCSCNWKEY, (caddr_t)&ifr) < 0)
			err(1, "SIOCSCNWKEY");
	}

	if (domain >= 0) {
		memset(&ifr, 0, sizeof(ifr));
		strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
		ifr.ifr_domain = domain;
		if (ioctl(skt, SIOCSCNWDOMAIN, (caddr_t)&ifr) < 0)
			err(1, "SIOCSCNWDOMAIN");
	}

	if (sflag == 0 && Sflag == 0)
		exit (0);

	if (Sflag) {
        	memset(&cnws, 0, sizeof(cnws));
		strncpy(cnws.ifr.ifr_name, interface,
		    sizeof(cnws.ifr.ifr_name));
		if (ioctl(skt, SIOCGCNWSTATUS, (caddr_t)&cnws) < 0)
			err(1, "SIOCGCNWSTATUS");
	}

	if (sflag) {
		memset(&cnwis, 0, sizeof(cnwis));
		strncpy(cnwis.ifr.ifr_name, interface,
		    sizeof(cnwis.ifr.ifr_name));
		if (ioctl(skt, SIOCGCNWSTATS, (caddr_t)&cnwis) < 0)
			err(1, "SIOCGCNWSTATS");
		cnwis.stats.nws_txretries[0] = 0;

		for (i = 1; i < 16; ++i)
			cnwis.stats.nws_txretries[0] +=
			    cnwis.stats.nws_txretries[i] * i;
	}

	if (rate == 0 && sflag) {
		memset(&ifr, 0, sizeof(ifr));
		strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
		if (ioctl(skt, SIOCGCNWDOMAIN, (caddr_t)&ifr) < 0)
			err(1, "SIOCGCNWDOMAIN");
		printf("%s:\n     0x%03x domain\n", interface, ifr.ifr_domain);
			printf("%10d rx\n", cnwis.stats.nws_rx);
		printf("%10d rxoverflow\n", cnwis.stats.nws_rxoverflow);
		printf("%10d rxoverrun\n", cnwis.stats.nws_rxoverrun);
		printf("%10d rxcrcerror\n", cnwis.stats.nws_rxcrcerror);
		printf("%10d rxframe\n", cnwis.stats.nws_rxframe);
		printf("%10d rxerrors\n", cnwis.stats.nws_rxerrors);
		printf("%10d rxavail\n", cnwis.stats.nws_rxavail);

		printf("%10d tx\n", cnwis.stats.nws_tx);
		printf("%10d txokay\n", cnwis.stats.nws_txokay);
		printf("%10d txabort\n", cnwis.stats.nws_txabort);
		printf("%10d txlostcd\n", cnwis.stats.nws_txlostcd);
		printf("%10d txerrors\n", cnwis.stats.nws_txerrors);
		printf("%10d txretries\n", cnwis.stats.nws_txretries[0]);
		for (i = 1; i < 16; ++i)
			if (cnwis.stats.nws_txretries[i])
				printf("%10s %10d %dx retries\n", "",
				    cnwis.stats.nws_txretries[i], i);
	}

	if (rate == 0 && Sflag) {
		printf("      0x%02x link integrity field\n",
		    cnws.data[0x4e]);
		printf("      0x%02x connection quality\n",
		    cnws.data[0x54]);
		printf("      0x%02x spu\n",
		    cnws.data[0x55]);
		printf("      0x%02x link quality\n",
		    cnws.data[0x56]);
		printf("      0x%02x hhc\n",
		    cnws.data[0x58]);
		printf("      0x%02x mhs\n",
		    cnws.data[0x6b]);
		printf(" %04x %04x revision\n",
		    *(u_short *)&cnws.data[0x66], *(u_short *)&cnws.data[0x68]);
		printf("        %c%c id\n",
		    cnws.data[0x6e], cnws.data[0x6f]);
	}

	if (rate == 0)
		exit (0);

	if (ioctl(0, TIOCGWINSZ, &ts) < 0)
		ts.ts_lines = 24;
	c = 0;

	if (Sflag) for (;;) {
		if (c-- == 0) {
			printf("lif  cq spu  lq hhc mhs\n");
			c = ts.ts_lines - 3;
		}
		printf(" %02x  %02x  %02x  %02x  %02x  %02x\n",
                    cnws.data[0x4e],
                    cnws.data[0x54],
                    cnws.data[0x55],
                    cnws.data[0x56],
                    cnws.data[0x58],
		    cnws.data[0x6b]);
		fflush(stdout);
		if (ioctl(skt, SIOCGCNWSTATUS, (caddr_t)&cnws) < 0)
			err(1, "SIOCGCNWSTATUS");
		sleep (rate);
	}

	for (;;) {
		if (c-- == 0) {
			printf("%10s %10s %10s %10s %10s %10s\n",
			    "tx-request", "tx-okay", "tx-error", "tx-retry",
			    "rx", "rx-error");
			c = ts.ts_lines - 3;
			memset(&onwis, 0, sizeof(onwis));
		}
		printf("%10d ", cnwis.stats.nws_tx - onwis.stats.nws_tx);
		printf("%10d ", cnwis.stats.nws_txokay - onwis.stats.nws_txokay);
		printf("%10d ", cnwis.stats.nws_txerrors - onwis.stats.nws_txerrors);
		printf("%10d ", cnwis.stats.nws_txretries[0] - onwis.stats.nws_txretries[0]);
		printf("%10d ", cnwis.stats.nws_rx - onwis.stats.nws_rx);
		printf("%10d\n", cnwis.stats.nws_rxerrors - onwis.stats.nws_rxerrors);
		fflush(stdout);
		sleep (rate);
		onwis = cnwis;

		if (ioctl(skt, SIOCGCNWSTATS, (caddr_t)&cnwis) < 0)
			err(1, "SIOCGCNWSTATS");
		cnwis.stats.nws_txretries[0] = 0;
		for (i = 1; i < 16; ++i)
			cnwis.stats.nws_txretries[0] +=
			    cnwis.stats.nws_txretries[i] * i;
	}
}
